examples/ml/raytracer.effekt

```diff
@@ -0,0 +1,275 @@
+
+
+def length(a: Vector): Double = dot(a, a).sqrt
+
+def unit(a: Vector): Vector = a.scale(1.0 / a.length)
+
+def reflectThrough(a: Vector, normal: Vector): Vector =
+  normal.scale(dot(a, normal)).scale(2.0) - a
+
+
+
+// this should be moved to the standard library
+
+def min(x: Double, y: Double): Double = if (x < y) x else y
+
+
+type Point = Vector
+def Point(x: Double, y: Double, z: Double) = Vector(x, y, z)
+
+
+// 0 -- 255
+record Color(red: Int, green: Int, blue: Int)
+
+// for precision, we are working with double values between 0.0 and 255.0
+// we only translate back at the very end.
+type ColorV = Vector
+val BLACK = Vector(0.0, 0.0, 0.0)
+val WHITE = Vector(255.0, 255.0, 255.0)
+
+// we perform color computation on vectors, however
+def toVector(c: Color): ColorV = Vector(c.red.toDouble, c.green.toDouble, c.blue.toDouble)
+
+// TODO here we should check for the range of 0-255
+def toColor(c: ColorV): Color = Color(c.x.toInt, c.y.toInt, c.z.toInt)
+
+type Light = Point
+
+// The unit interval 0.0 - 1.0
+type Percent = Double
+
+record Camera(position: Point, fieldOfView: Double, direction: Vector)
+
+// for now all objects are spheres
+record Material(
+  color: Color,
+  specular: Percent,
+  lambert: Percent,
+  ambient: Percent
+)
+record SceneObject(name: String, position: Point, radius: Double, material: Material)
+val DUMMY = SceneObject("DUMMY", Point(0.0, 0.0, 0.0), 0.0, Material(Color(0, 0, 0), 0.0, 0.0, 0.0))
+
+record Scene(width: Int, height: Int, camera: Camera, lights: List[Light], objects: List[SceneObject])
+
+
+record Intersection(distance: Double, pointAtTime: Point, normal: Vector, object: SceneObject)
+
+
+def sphereNormal(sphere: SceneObject, surface: Point) = (surface - sphere.position).unit
+
+def intersectSphere(pos: Point, dir: Vector, sphere: SceneObject): Option[Double] = {
+  val distance = sphere.position - pos;
+  val v = dot(distance, dir);
+  val discriminant = (sphere.radius * sphere.radius) - dot(distance, distance) + (v * v)
+  if (discriminant < 0.0) { None() } else {
+    Some(v - sqrt(discriminant))
+  }
+}
+
+def intersectScene(pos: Point, dir: Vector, scene: Scene): Option[Intersection] = {
+
+  // TODO this is still way to Javascripty
+  var bestDist in global = 0.0;
+  var bestObj in global = DUMMY;
+  var found in global = false;
+
+  scene.objects.foreach { sphere =>
+    intersectSphere(pos, dir, sphere).foreach { dist =>
+      if (not(found) || dist < bestDist) {
+        bestDist = dist;
+        bestObj = sphere;
+        found = true
+      } else {  () }
+    }
+  };
+
+  if (found) {
+    val pointAtTime = pos + dir.scale(bestDist);
+    val normal = sphereNormal(bestObj, pointAtTime);
+    Some(Intersection(bestDist, pointAtTime, normal, bestObj))
+  } else { None() }
+}
+
+
+def trace(pos: Point, dir: Vector, scene: Scene, depth: Int): ColorV = {
+
+  def surface(pos: Point, dir: Vector, scene: Scene, depth: Int, intersection: Intersection): ColorV = {
```

Comment:
Compared to the original raytracer example in `examples/pos` I had to manually reorder (and partially re-nest) the function definitions in order to be sorted topologically. @JonathanStarup Could you maybe open up a ticket so that we remember implementing this as a preprocessing step before emitting ML?

Reply:
I also rewrite the List stdlib to use tailrecursive functions instead of local mutable state since we do not know how fast the latter is in ML at the moment.